Add template for accepting vector diffusivity in diffusive flux postprocessor

GiudGiud opened the pull request:
I only did the SideDiffusiveFlux one, and only RealVectorValue. 

Once this is validated, we can consider adding the tensor already and the interface version of this. 

I kind of wish I could have defined a partial specialization of the product member function, but I could not figure it out. Someone better than me at template meta-programming could let me know how to do that cleanly (eg not partially specializing the whole class first)

GiudGiud committed:
Add templated vector diffusivity (should be a tensor)

GiudGiud committed:
Add support for RealVectorValue diffusivities in diffusiveFlux posptproc, refs #18118

GiudGiud committed:
Add tests for side diffusive flux integral, both scalar and vector diffusivities, refs #18118
Remove odd cast from RealVectorValue to Real and back
Move ad_indexing limitation in tests to different test nesting level

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/761789/) on d7061e7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18119/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18119/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9db85613d2e880041b998b69601e95a9ee0b2c31`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/761818/) on 931b923 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18119/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18119/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9db85613d2e880041b998b69601e95a9ee0b2c31`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/761923/) on 8733064 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18119/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/771319/) on 7f169df wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18119/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18119/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f59e5424802e6b294e602ea4597ce6e78bfa08ed`

GiudGiud commented on framework/src/postprocessors/SideDiffusiveFluxIntegral.C:
```diff
@@ -60,11 +62,36 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
-    return -MetaPhysicL::raw_value(_diffusion_coef[_qp]) * _grad_u[_qp] * _normals[_qp];
+    return -diffusivity_gradient_product(_grad_u[_qp],
+                                         MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
+           _normals[_qp];
```
I've tried changing this a couple times, clang won't let me

lindsayad replied:
What is the error?

GiudGiud replied:
precheck whining

GiudGiud commented:
Internal app test failures are unrelated, though that test does use this PP.

lindsayad commented on framework/include/postprocessors/SideDiffusiveFluxIntegral.h:
```diff
@@ -36,5 +38,12 @@ class SideDiffusiveFluxIntegralTempl : public SideIntegralVariablePostprocessor
+  const GenericMaterialProperty<T, is_ad> & _diffusion_coef;
+
+private:
+  /// Routine to get the diffusive flux with a Real diffusivity
+  RealVectorValue diffusivity_gradient_product(RealVectorValue grad_u, Real diffusivity);
+
+  /// Routine to get the diffusive flux with a RealVectorValue diffusivity
+  RealVectorValue diffusivity_gradient_product(RealVectorValue grad_u, RealVectorValue diffusivity);
```
MOOSE style is camel case, so `diffusivityGradientProduct`

lindsayad commented on framework/src/postprocessors/SideDiffusiveFluxIntegral.C:
```diff
@@ -60,11 +62,36 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
+                                                                       RealVectorValue diffusivity)
+{
+  RealVectorValue d_grad_u = grad_u;
+  for (unsigned int i = 0; i < LIBMESH_DIM; i++)
+    d_grad_u *= diffusivity(i);
```
This is not a component-wise multiplication. Isn't that what you want? What you are doing now will yield this vector:

(grad_u0 * d0 * d1 * d2, grad_u1 * d0 * d1 * d2, grad_u2 * d0 * d1 * d2)

GiudGiud replied:
good catch. Should be d_grad_u(i)

GiudGiud committed:
Address Alex's review:
-fix bug with vector diffusivities
-make the product function camelcase
Add test for anisotropic vector diffusivities to be able to catch said bug

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/772716/) on 5c03b7d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18119/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18119/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f59e5424802e6b294e602ea4597ce6e78bfa08ed`

lindsayad commented on framework/include/postprocessors/SideDiffusiveFluxIntegral.h:
```diff
@@ -36,5 +38,12 @@ class SideDiffusiveFluxIntegralTempl : public SideIntegralVariablePostprocessor
+
+private:
+  /// Routine to get the diffusive flux with a Real diffusivity
+  RealVectorValue DiffusivityGradientProduct(RealVectorValue grad_u, Real diffusivity);
+
+  /// Routine to get the diffusive flux with a RealVectorValue diffusivity
+  RealVectorValue DiffusivityGradientProduct(RealVectorValue grad_u, RealVectorValue diffusivity);
```
```suggestion
  RealVectorValue diffusivityGradientProduct(const RealVectorValue & grad_u, Real diffusivity);

  /// Routine to get the diffusive flux with a RealVectorValue diffusivity
  RealVectorValue diffusivityGradientProduct(const RealVectorValue & grad_u, const RealVectorValue & diffusivity);
```

lindsayad commented on framework/src/postprocessors/SideDiffusiveFluxIntegral.C:
```diff
@@ -60,11 +62,35 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
 
     // FIXME Get the diffusion coefficient on the face, see #16809
-    return -MetaPhysicL::raw_value(_diffusion_coef[_qp]) * grad_u * _normals[_qp];
+    return -DiffusivityGradientProduct(grad_u, MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
```
```suggestion
    return -diffusivityGradientProduct(grad_u, MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
```

lindsayad commented on framework/src/postprocessors/SideDiffusiveFluxIntegral.C:
```diff
@@ -60,11 +62,35 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
   }
   else
-    return -MetaPhysicL::raw_value(_diffusion_coef[_qp]) * _grad_u[_qp] * _normals[_qp];
+    return -DiffusivityGradientProduct(_grad_u[_qp], MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
```
```suggestion
    return -diffusivityGradientProduct(_grad_u[_qp], MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
```

lindsayad commented on framework/src/postprocessors/SideDiffusiveFluxIntegral.C:
```diff
@@ -60,11 +62,35 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
-template class SideDiffusiveFluxIntegralTempl<true>;
+template <bool is_ad, typename T>
+RealVectorValue
+SideDiffusiveFluxIntegralTempl<is_ad, T>::DiffusivityGradientProduct(RealVectorValue grad_u,
+                                                                     Real diffusivity)
+{
+  return grad_u * diffusivity;
+}
+
+template <bool is_ad, typename T>
+RealVectorValue
+SideDiffusiveFluxIntegralTempl<is_ad, T>::DiffusivityGradientProduct(RealVectorValue grad_u,
+                                                                     RealVectorValue diffusivity)
```
```suggestion
SideDiffusiveFluxIntegralTempl<is_ad, T>::diffusivityGradientProduct(const RealVectorValue & grad_u,
                                                                     const Real diffusivity)
{
  return grad_u * diffusivity;
}

template <bool is_ad, typename T>
RealVectorValue
SideDiffusiveFluxIntegralTempl<is_ad, T>::diffusivityGradientProduct(const RealVectorValue & grad_u,
                                                                     const RealVectorValue & diffusivity)
```

lindsayad commented on test/tests/postprocessors/side_diffusive_flux_integral/tests:
```diff
@@ -0,0 +1,40 @@
+      input = 'side_diffusive_flux_integral.i'
+      exodiff = 'side_diffusive_flux_integral_vector.e'
+      cli_args = "Outputs/file_base=side_diffusive_flux_integral_vector Postprocessors/avg_flux_right/type=SideVectorDiffusivityFluxIntegral Postprocessors/avg_flux_right/diffusivity=diffusivity_vec Postprocessors/inactive=''"
+      detail = 'and with an anisotropic vector diffusivity.'
```
with the way things read in the actual sqa document produced, I would just do
```suggestion
      detail = 'with an anisotropic vector diffusivity.'
```

lindsayad commented on test/tests/postprocessors/side_diffusive_flux_integral/tests:
```diff
@@ -0,0 +1,40 @@
+      exodiff = 'side_diffusive_flux_integral_fv_vector.e'
+      cli_args = "Outputs/file_base=side_diffusive_flux_integral_fv_vector Postprocessors/avg_flux_right/type=SideVectorDiffusivityFluxIntegral Postprocessors/avg_flux_right/diffusivity=diffusivity_vec Postprocessors/inactive=''"
+      ad_indexing_type = 'global'
+      detail = 'and with a vector diffusivity.'
```
```suggestion
      detail = 'with a vector diffusivity.'
```

GiudGiud committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/772806/) on 5c0b310 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18119/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18119/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f59e5424802e6b294e602ea4597ce6e78bfa08ed`
